### PR TITLE
Fold Const(fp) -> QuantizeLinear into Const(int)

### PR DESCRIPTION
**Pull request by marcblaaa**
Adds a constprop-onnx rewrite that folds Const(fp) -> QuantizeLinear into a pre-computed integer constant.

### CODE REVIEW

**Review by akj10 — APPROVED** [review]

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 1]
```diff
@@ -1691,6 +1691,118 @@ class RemoveQDQForConst : public OpRewritePattern<ONNXOp> {
...
+      if (!zpElems)
+        return rewriter.notifyMatchFailure(
+            qOp, "y_zero_point is not a constant");
+    }
+
+    // Output storage type must be a plain integer (skip float8).
+    Type yElemType = getElementTypeOrSelf(qOp.getY());
+    Type storageType = yElemType;
+    if (auto qType = dyn_cast<mlir::quant::QuantizedType>(yElemType))
+      storageType = qType.getStorageType();
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp]: is this really allowed/intended? Having a quantize op with a quantized type as ouput?

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 2]
```diff
@@ -1691,6 +1691,118 @@ class RemoveQDQForConst : public OpRewritePattern<ONNXOp> {
...
+    ElementsAttr scaleF64 =
+        broadcastToX(elementsBuilder.castToFPElementType(scaleElems, f64Type));
+    ElementsAttr scaled = elementsBuilder.combine(xF64, scaleF64, combinedType,
+        elementwiseBinaryOpCombiner<ONNXDivOp>(f64Type));
+
+    // rounded = round-to-nearest-ties-to-even(scaled).
+    ElementsAttr rounded =
+        elementsBuilder.transform(scaled, f64Type, [](WideNum n) {
+          return WideNum::widen<BType::DOUBLE>(
+              std::nearbyint(n.narrow<BType::DOUBLE>()));
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp]: std::nearbyint does not generally implement round-to-nearest-ties-to-even, it depends on the set rounding mode

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 3]
```diff
@@ -1691,6 +1691,118 @@ class RemoveQDQForConst : public OpRewritePattern<ONNXOp> {
...
+    auto intType = dyn_cast<IntegerType>(storageType);
+    if (!intType)
+      return rewriter.notifyMatchFailure(qOp, "non-integer output type");
+
+    auto xType = cast<ShapedType>(qOp.getX().getType());
+    if (!xType.hasStaticShape())
+      return rewriter.notifyMatchFailure(qOp, "x has dynamic shape");
+    int64_t xRank = xType.getRank();
+
+    FloatType f64Type = rewriter.getF64Type();
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp:1764]: you do everything on f64 type instead of the real float type.
this does not match the reference spec:
https://github.com/onnx/onnx/blob/f320feb96b88ad895541b7caaa63ea1c8000423e/onnx/reference/ops/op_quantize_linear.py#L87

and can lead to rounding errors, for example for 
 x = np.float32(-99999.85)
 scale = np.float32(0.1)  .
 
 (Generally onnx does the const folding on fp64, but i am not sure if this is the best decision. I will ask upstream why they did do it like that).
 
 I think doing it on fp64 should be fine for now, but if we get accuracy issues downstream we may need to update it to work on the real float type.

**Comment by ilango100** [comment]
I've implemented this pattern that uses llvm::APFloat and llvm::APInt here: #806 The full PR also implements Const -> DQ, when DQ is followed by dataflow ops.

File: src/Dialect/ONNX/Transforms/ConstProp.cpp (hunk 3 above)
Comment (jorickert) [reply src/Dialect/ONNX/Transforms/ConstProp.cpp:1764]: please create a (internal) follow-up ticket to use the exact float semantics

File: src/Dialect/ONNX/Transforms/ConstProp.cpp (hunk 1 above)
Comment (marcblaaa) [reply src/Dialect/ONNX/Transforms/ConstProp.cpp]: yes, the op definition does allow a quant_QuantizedType result, but our QDQ flow always returns a plain integer type with separate scale/zp operand so no need to do the unwrapping. Removed the unwrapping.

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 4]
```diff
@@ -1740,11 +1859,13 @@ void onnx_mlir::getConstPropONNXToONNXPatterns(
   patterns.insert<IfOfConst>(patterns.getContext());
   patterns.insert<LoopUnroll>(patterns.getContext());
   patterns.insert<ConstPropConcatFromSequence>(patterns.getContext());
-  if (enableQDQ)
+  if (enableQDQ) {
     patterns.add<RemoveQDQForConst<ONNXSliceOp>,
         RemoveQDQForConst<ONNXTransposeOp>, RemoveQDQForConst<ONNXReshapeOp>,
         RemoveQDQForConst<ONNXSqueezeOp>, RemoveQDQForConst<ONNXUnsqueezeOp>,
         RemoveQDQForConst<ONNXGatherOp>>(patterns.getContext());
+    patterns.add<ConstFoldQuantizeLinearOnConst>(patterns.getContext());
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp:1877]: Can we enable this by default ? Or maybe behind a different flag if you want a flag? I do not see cases where we would not like to use it

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 5]
```diff
@@ -1691,6 +1691,125 @@ class RemoveQDQForConst : public OpRewritePattern<ONNXOp> {
...
+  using OpRewritePattern<ONNXQuantizeLinearOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXQuantizeLinearOp qOp, PatternRewriter &rewriter) const override {
+    // Blocked quantization is out of scope.
+    if (qOp.getBlockSize() != 0)
+      return rewriter.notifyMatchFailure(qOp, "blocked quantization");
+
+    if (!qOp.getSaturate())
+      return rewriter.notifyMatchFailure(qOp, "only saturate=1 is supported");
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp]: nit: This check is not needed, saturate is only relevant for fp8 (i know ai recommended that)

File: src/Dialect/ONNX/Transforms/ConstProp.cpp [hunk 6]
```diff
@@ -1691,6 +1691,125 @@ class RemoveQDQForConst : public OpRewritePattern<ONNXOp> {
...
+      if (!zpElems)
+        return rewriter.notifyMatchFailure(
+            qOp, "y_zero_point is not a constant");
+    }
+
+    auto intType = dyn_cast<IntegerType>(getElementTypeOrSelf(qOp.getY()));
+    if (!intType)
+      return rewriter.notifyMatchFailure(qOp, "non-integer output type");
+    // castToIntElementType computes the saturation range with 64-bit shifts.
+    if (intType.getWidth() > 64)
```
Comment (jorickert) [inline src/Dialect/ONNX/Transforms/ConstProp.cpp]: nit:  This check is not needed,onnx does not have int types with a width higher than 64 bits

**Review by jorickert — APPROVED** [review]
Looks good, thanks.

Could you maybe extend the tests a bit to test edge cases?
For example with quant_saturates_uint8_upper: x = [260.0], scale = 1.0, zp = ui8 0, expect dense<[255]> : tensor<1xui8>, not wrap to 4.

quant_saturates_uint8_after_zp_add: x = [250.0], scale = 1.0, zp = ui8 10, expect 255.

quant_saturates_int8_lower: x = [-200.0], scale = 1.0, zp = i8 0, expect -128.

quant_rounds_ties_to_even: x = [2.5, 3.5, -2.5, -3.5], scale = 1.0, zp = i8 0, expect [2, 4, -2, -4].

File: src/Dialect/ONNX/Transforms/ConstProp.cpp (hunk 5 above)
Comment (marcblaaa) [reply src/Dialect/ONNX/Transforms/ConstProp.cpp]: Thanks, I also was not sure about this change.

File: src/Dialect/ONNX/Transforms/ConstProp.cpp (hunk 4 above)
Comment (marcblaaa) [reply src/Dialect/ONNX/Transforms/ConstProp.cpp:1877]: Enabled by default now

**Comment by akj10** [comment]
@marcblaaa what is pending here?

**Review by jorickert — APPROVED** [review]
